src/clr-angular/data/datagrid/providers/column-order-model.service.ts

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { Injectable } from '@angular/core';
+import { ColumnOrdersCoordinatorService } from './column-orders-coordinator.service';
+import { DatagridHideableColumnModel } from '../datagrid-hideable-column.model';
+import { DomAdapter } from '../../../utils/dom-adapter/dom-adapter';
+
+/**
+ * This is a model service that's responsible for:
+ * 1. Sharing the flex order data between datagrid-column and reorderabble-droppable
+ * 2. Updating the flex order data and tells ColumnOrderCoordinatorService when it should broadcast
+ * 3. Returning models at the previous and next flex orders
+ */
+
+@Injectable()
+export class ColumnOrderModelService {
+  constructor(private columnOrderCoordinatorService: ColumnOrdersCoordinatorService, private domAdapter: DomAdapter) {}
+
+  public flexOrder: number;
+
+  public headerEl: any;
+
+  public hideableColumnModel: DatagridHideableColumnModel;
+
+  get columnGroupId() {
+    return this.columnOrderCoordinatorService.columnGroupId;
+  }
+
+  get isAtFirst(): boolean {
+    return this.flexOrder === 0;
+  }
+
+  get isAtEnd(): boolean {
+    return this.flexOrder === this.columnOrderCoordinatorService.orderModels.length - 1;
```

Comment:
What if the last column is hidden? Then the one before last is visually the last one. Shouldn't `isAtEnd` account for that?

Reply:
I have changed the `isAtEnd` property to `isLast` in the next PR. I have also wrote another property called `isLastVisible` that basically addresses this issue you are raising here. And this property needed to figure out whether it needs to project a separator in a column header or not (as we know the last visible header shouldn't have a separator in itself).  I haven't really needed to implement `isFirstVisible` as I haven't faced a need to do so 🤔